### PR TITLE
Fix series comparison operators when dealing with zero rank numpy arrays

### LINKED ISSUE #13006: BUG: Type error when comparing numpy scalar to pandas series
Hi,

I'm getting strange results when comparing a numpy float to a pandas series. It seems to think that the numpy float is some sort of array...

```python
import numpy as np
import pandas as pd
np.float64(0) < pd.Series([1,2,3],dtype=np.float64)
```
#### Expected Output
```
0    True
1    True
2    True
dtype: bool
```
#### Actual Output
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-0803a1472416> in <module>()
      1 import numpy as np
      2 import pandas as pd
----> 3 np.float64(0) < pd.Series([1,2,3],dtype=np.float64)
        global np.float64 = <type 'numpy.float64'>
        global pd.Series = <class 'pandas.core.series.Series'>
        global dtype = undefined

/usr/local/lib/python2.7/dist-packages/pandas/core/ops.pyc in wrapper(self=0    1.0
1    2.0
2    3.0
dtype: float64, other=array(0.0), axis=None)
    737             return NotImplemented
    738         elif isinstance(other, (np.ndarray, pd.Index)):
--> 739             if len(self) != len(other):
        global len = undefined
        self = 0    1.0
1    2.0
2    3.0
dtype: float64
        other = array(0.0)
    740                 raise ValueError('Lengths must match to compare')
    741             return self._constructor(na_op(self.values, np.asarray(other)),

TypeError: len() of unsized object


```
#### output of ``pd.show_versions()``
```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 4.3.0-040300-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: None
pip: 8.1.1
setuptools: 20.3.1
Cython: None
numpy: 1.11.0
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.0.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2015.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: 0.8
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
```

### PR DESCRIPTION
**Pull request by gliptak**
- [x] closes #13006
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gliptak** [comment]
http://stackoverflow.com/questions/4565749/how-detect-length-of-a-numpy-array-with-only-one-element#4565910

Partially based on @sldion 's changes. Thanks

File: pandas/core/ops.py [hunk 1]
```diff
@@ -754,7 +754,8 @@ def wrapper(self, other, axis=None):
         elif isinstance(other, pd.DataFrame):  # pragma: no cover
             return NotImplemented
         elif isinstance(other, (np.ndarray, pd.Index)):
-            if len(self) != len(other):
+            lo = len(np.atleast_1d(other))
+            if lo != 1 and len(self) != lo:
```
Comment (jreback) [inline pandas/core/ops.py]: why are you adding this check for != 1? this seems to be hiding thing. Further why don't you do:

``other = np.atleast_1d(other)`` and be done?

Comment (gliptak) [inline pandas/core/ops.py]: ```len(other) == 1``` shouldn't throw the exception, as ```other``` is broadcasted by ```_constructor``` ...

Comment (jreback) [inline pandas/core/ops.py]: much better to fix it than put this kind of conditional in

Comment (gliptak) [inline pandas/core/ops.py]: how do you see this changed differently?

Comment (jreback) [inline pandas/core/ops.py]: exactly as I just said

Comment (gliptak) [inline pandas/core/ops.py]: I tried, but that fails the (new) testcase

Would this work?
```
            try:
                return self._constructor(na_op(self.values, np.asarray(other)),
                                         index=self.index).__finalize__(self)
            except:
                raise ValueError('Lengths must match to compare')
```

Comment (jreback) [inline pandas/core/ops.py]: no that's not a good thing to do, we don't like non-specific try:excepts.

you can specifically deal with it by something like:

``if lib.isscalar(lib.item_from_zerodim(other)):``

Comment (gliptak) [inline pandas/core/ops.py]: modified with this above

File: pandas/core/ops.py [hunk 2]
```diff
@@ -754,8 +754,9 @@ def wrapper(self, other, axis=None):
         elif isinstance(other, pd.DataFrame):  # pragma: no cover
             return NotImplemented
         elif isinstance(other, (np.ndarray, pd.Index)):
-            if len(self) != len(other):
-                raise ValueError('Lengths must match to compare')
+            if not lib.isscalar(lib.item_from_zerodim(other)):
+                if len(self) != len(other):
```
Comment (jreback) [inline pandas/core/ops.py]: still to much nested if statements
maybe just try to convert a 0-dim ndarray to a scaler instead

Comment (gliptak) [inline pandas/core/ops.py]: there have to be two if statements, to determine scalar or not, and validate length being the same

File: doc/source/whatsnew/v0.18.2.txt [hunk 3]
```diff
@@ -315,6 +315,6 @@ Bug Fixes
 - Bug in ``groupby`` where ``apply`` returns different result depending on whether first result is ``None`` or not (:issue:`12824`)
 
 
-
+- Bug in series comparison operators when dealing with zero rank numpy arrays (:issue:`13006`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: Use ``Series`` (surrounded by double-ticks) instead of "series" and change "numpy" to "NumPy"

Comment (gliptak) [inline doc/source/whatsnew/v0.18.2.txt]: updated

File: doc/source/whatsnew/v0.18.2.txt [hunk 4]
```diff
@@ -98,7 +98,7 @@ API changes
 
 
 - Non-convertible dates in an excel date column will be returned without conversion and the column will be ``object`` dtype, rather than raising an exception  (:issue:`10001`)
-- An ``UnsupportedFunctionCall`` error is now raised if numpy ufuncs like ``np.mean`` are called on groupby or resample objects (:issue:`12811`)
+- An ``UnsupportedFunctionCall`` error is now raised if NumPy ufuncs like ``np.mean`` are called on groupby or resample objects (:issue:`12811`)
 - Calls to ``.sample()`` will respect the random seed set via ``numpy.random.seed(n)`` (:issue:`13161`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: Ha, ha!  Good catch, as that was one of my own changes :smile:

File: pandas/core/ops.py (hunk 2 above)
Comment (gliptak) [inline pandas/core/ops.py]: @jreback Would you like this to be changed?

**Comment by gliptak** [comment]
@jreback Are you good with this PR? Thanks

File: pandas/tests/series/test_operators.py [hunk 5]
```diff
@@ -264,6 +264,15 @@ def test_operators_timedelta64(self):
         rs[2] += np.timedelta64(timedelta(minutes=5, seconds=1))
         self.assertEqual(rs[2], value)
 
+    def test_operator_series_comparison_zerorank(self):
+        self.assert_series_equal((np.float64(0) > pd.Series([1, 2, 3])), (0.0 >
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: add the issue number as a comment

File: pandas/tests/series/test_operators.py [hunk 6]
```diff
@@ -264,6 +264,15 @@ def test_operators_timedelta64(self):
...
+                                 pd.Series([1, 2, 3])))
+        self.assert_series_equal((pd.Series([1, 2, 3]) < np.float64(0)),
```
Comment (jreback) [inline pandas/tests/series/test_operators.py]: use the 

```
result = 
expected = 
assert_series_equal...
```
format

File: pandas/tests/series/test_operators.py (hunk 5 above)
Comment (gliptak) [inline pandas/tests/series/test_operators.py]: added

File: pandas/tests/series/test_operators.py (hunk 6 above)
Comment (gliptak) [inline pandas/tests/series/test_operators.py]: updated

File: pandas/core/ops.py (hunk 2 above)
Comment (jreback) [inline pandas/core/ops.py]: you can do this in an ``and``. put a comment that need to care for zero-dim ndarrays

**Comment by gliptak** [comment]
build is green

File: pandas/core/ops.py [hunk 7]
```diff
@@ -754,7 +754,9 @@ def wrapper(self, other, axis=None):
         elif isinstance(other, pd.DataFrame):  # pragma: no cover
             return NotImplemented
         elif isinstance(other, (np.ndarray, pd.Index)):
-            if len(self) != len(other):
+            # do not check length of zerorank array
+            if not lib.isscalar(lib.item_from_zerodim(other)) and \
+               len(self) != len(other):
                 raise ValueError('Lengths must match to compare')
```
Comment (gfyoung) [inline pandas/core/ops.py]: I've been nitpicked on this before, but @jreback is there a preference for backslash or parens for multi-line conditionals?

File: pandas/core/ops.py [hunk 8]
```diff
@@ -754,7 +754,9 @@ def wrapper(self, other, axis=None):
         elif isinstance(other, pd.DataFrame):  # pragma: no cover
             return NotImplemented
         elif isinstance(other, (np.ndarray, pd.Index)):
-            if len(self) != len(other):
+            # do not check length of zerorank array
+            if not lib.isscalar(lib.item_from_zerodim(other)) and \
```
Comment (gfyoung) [inline pandas/core/ops.py]: Nitpicking: I think it's "zero rank" instead of "zerorank"

File: pandas/core/ops.py (hunk 7 above)
Comment (jreback) [inline pandas/core/ops.py]: either one are fine really. I guess in this case parens might be nicer. I'll fix on the merge

**Comment by jreback** [comment]
thanks @gliptak

**Pull request closed without merging**